File: java-to-go-projects/16-cron-parser/Main.java
package cron;

import java.time.LocalDateTime;
import java.util.*;

/**
 * CLI entry point for the cron expression parser.
 *
 * Usage: java Main "*/15 0 1,15 * 1-5" [--next=5]
 *
 * Go structure:
 *   main.go                  — CLI parsing, display
 *   cron/schedule.go         — Schedule struct, PrettyPrint
 *   cron/parser.go           — Parse(expr) → Schedule, parseField with range/step/list
 *   cron/next_time.go        — NextN(schedule, from, count), Matches(schedule, time)
 *                               Uses time.Time, time.Add(time.Minute)
 *
 * Rust structure:
 *   main.rs                  — CLI parsing
 *   schedule.rs              — Schedule struct, Display impl, HashSet<u32> for fields
 *   parser.rs                — parse(expr) → Schedule, parse_field with split/parse
 *   next_time.rs             — next_n(schedule, from, count) using chrono::NaiveDateTime
 *                               chrono crate for date/time manipulation
 *
 * Key learning:
 *   - Go: strings.Split, strconv.Atoi, time package, map[int]bool or bitset for fields
 *   - Rust: str::split, parse::<u32>(), HashSet, chrono crate
 */
public class Main {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: cron-parser <expression> [--next=5]");
            System.err.println("Example: cron-parser \"*/15 0 1,15 * 1-5\"");
            System.exit(1);
        }

        String expression = args[0];
        int nextCount = 5;

        for (int i = 1; i < args.length; i++) {
            if (args[i].startsWith("--next=")) {
                nextCount = Integer.parseInt(args[i].split("=")[1]);
            }
        }

        try {
            Schedule schedule = Parser.parse(expression);

            System.out.println("Cron: " + expression);
            System.out.println();
            System.out.println(schedule.prettyPrint());

            System.out.println();
            System.out.println("Next " + nextCount + " execution times:");
            List<LocalDateTime> times = NextTime.nextN(schedule, LocalDateTime.now(), nextCount);
            for (LocalDateTime t : times) {
                System.out.println("  " + t);
            }

        } catch (IllegalArgumentException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: java-to-go-projects/16-cron-parser/go/main.go
package cronparser
package main

import (
	"fmt"
	"os"
	"strings"
)

type Schedule struct {
	Minutes     map[int]bool // 0-59
	Hours       map[int]bool // 0-23
	DaysOfMonth map[int]bool // 1-31
	Months      map[int]bool // 1-12
	DaysOfWeek  map[int]bool // 0-6 (Sun=0)
	Original    string
}

func (s Schedule) String() string {
	panic("todo")
}

// parse parses a 5-field cron expression (or named shortcut) into a Schedule.
// Supports: *, ranges (N-M), steps (*/S, N-M/S), lists (N,M,O).
// Named shortcuts: @yearly, @monthly, @weekly, @daily, @hourly.
func parse(expression string) (*Schedule, error) {
	_ = strings.Fields
	panic("todo")
}

// parseField parses a single cron field into a set of values within [min, max].
func parseField(field string, min, max int) (map[int]bool, error) {
	panic("todo")
}

// matches checks if a given date/time matches the schedule.
func matches(s *Schedule, minute, hour, day, month, weekday int) bool {
	panic("todo")
}

// nextN calculates the next N execution times from a starting point.
// Iterates minute-by-minute from start.
func nextN(s *Schedule, year, month, day, hour, minute, count int) [][5]int {
	panic("todo")
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <cron-expression> [--next=5]\n", os.Args[0])
		os.Exit(1)
	}

	panic("todo: parse cron expression, pretty-print schedule, show next N times")
}
